### ISSUE TITLE
Document ways to properly authenticate only some methods in a controller

### ISSUE THREAD

**Issue by shimks** [strongloop/loopback-next#1334]
## Description / Steps to reproduce / Feature proposal
Copying from #1275 
> Just got back from a talk with @raymondfeng and told me what is happening here. Each time a route is accessed, a controller instance is created and resolves the injections given to it (in this case it'd be `CURRENT_USER`). When routes without `@authenticate` decoration are accessed, injections still try to be resolved, and in this case they will error out since `CURRENT_USER` hasn't been bound by the authentication function in the sequence.

> Here are three approaches to get this working in the intended way:
> 1. Pass in `{optional: true}` in `@inject` (`@inject(AuthenticationBindings.CURRENT_USER, {optional: true})`)
This allows injection to not error out even if `CURRENT_USER` is undefined
> 2. Inject `CURRENT_USER` as a method argument to operations requiring authentication
This allows the injection to only occur when the authenticated methods are being invoked
> 3. Use `@inject.getter` to retrieve the injection as the method is being invoked
This allows injection to be resolved when the method is being invoked and not when the controller is being instantiated.

There should be documentation for handling this use case in the README

_See [Reporting Issues](http://loopback.io/doc/en/contrib/Reporting-issues.html) for more tips on writing good issues_

**Comment by virkt25** [comment]
+1 for a more helpful error message if we throw one (should be improved anyways in case someone accidentally went down this path). 

I like the idea of injecting `CURRENT_USER` as a method argument / having it resolved as a getter (but with this approach I feel it's more complicated as a user would need to know they need to do this).

**Comment by jannyHou** [comment]
By injecting the `CURRENT_USER` when instantiate the controller class, each call has a consistent user if the invoked method calls other methods that need authentication. 

While by using `@inject.getter`, does it still guarantee the user retrieved in a chain of methods is always the same?

**Comment by jannyHou** [comment]
> Treat CURRENT_USER as a binding that's always defined, but provides undefined value when the user is not authenticated. Pro: it's easy to receive the current user from the context. Con: the type becomes UserProfile | undefined, consumers must be prepared to handle the case when the user is not defined.

Maybe we can define a default value for `UserProfile` that represents an anonymous user, then we don't need a union type to handle `undefined`...

**Comment by bajtos** [comment]
IIRC the constraints of our current authentication design, I think the following should be easy to achieve:

```ts
// Use the default strategy configured elsewhere
@authenticate()
@get('/whoami')
whoAmI(
  @injectCurrentUser() user: UserProfile
) {}

// Use a custom strategy
@authenticate('BasicStrategy')
@get('/whoami')
whoAmI(
  @injectCurrentUser() user: UserProfile
) {}
```

We can even shorten `@injectCurrentUser` to `@currentUser` or similar.

**Comment by joeytwiddle** [comment]
An afterthought: Would it be possible for `@authenticate()` or for `@injectCurrentUser()` to set `this.user`?

That might reduce the boilerplate, if we don't have to keep accepting the user object.

And it would also achieve parity with the current `@inject(AuthenticationBindings.CURRENT_USER)` in case we decide to keep that for developers who _do_ want authentication on every route.

**Comment by bajtos** [comment]
> An afterthought: Would it be possible for @authenticate() or for @injectCurrentUser() to set this.user?


You can inject the current user in your controller constructor. Depending on how we decide to handle CURRENT_USER when no user is authenticated (throw an error or return `undefined`), it is probably safer to inject a getter instead.

```ts
class UserController {
  constructor(
    @inject.getter(AuthenticationBindings.CURRENT_USER) 
    public getCurrentUser: Getter<UserProfile>,
  ) {}

  @authenticate()
  @get('/whoami')
  whoAmI() {
     console.log('I am ', this.getCurrentUser());
  }
}
```

**Comment by bajtos** [comment]
Authentication is out of scope of 4.0 GA, removing this story from the release.

**Comment by dhmlau** [comment]
@emonddr , would this be covered by your recently created task to clean up the docs? https://github.com/strongloop/loopback-next/issues/2940

**Comment by emonddr** [comment]
@dhmlau , we can definitely have a section on this. I've only ever decorate a controller method with `authenticate(name,options)` IF I want to create an authenticated endpoint, and I've only ever injected the `current user` on a controller method as a parameter on an authenticated endpoint (never in a controller constructor), but I will discuss with @jannyHou  and @raymondfeng . :)

**Comment by hazimdikenli** [comment]
I think you should use authentication decorator/injector either on class levels, or on method levels, does not make sense trying to support both options at the same time. If there is no way to prevent this decorator/injector from being used on class or method levels, let the developer handle the problems.

Like a controller might have only one method which needs authentication, let them use authentication decorator on method level. But if they use it on the class level and one of the methods does not need authentication, they should refactor their code, 

1.   either by adding another controller to house that method ; like splitting a users controller to have membership services, self sign-in services, or password recovery services etc. 
2. or by removing authentication decorator from class level and applying it on method levels.

Considering that this will be used in parallel with authorization decorators (or guards) on methods, what you'll end up doing is, add an authenticaton decorator at the class (controller level) so you have access to the current user and you are preventing anonymous access, and if required, add an authorization decorator on the method level, which requires a current user to check against the specified privileges on the decorator.

**Comment by dhmlau** [comment]
Discussion with @raymondfeng @jannyHou @emonddr:
- To answer the original question, `@authenticate` decorator is applicable at the method level.   
- If we want to allow decorators to be used at class level and method level can inherit from the class level, we can use interceptor. 

### Acceptance Criteria
- [ ] Document how to use the `@authenticate` decorator so that it is applicable at the method level (which is the current usage anyway).

**Comment by dhmlau** [comment]
@emonddr , per the discussion in the estimation meeting, could you please review the original description and see if your docs PR is going to cover that? Thanks.

**Comment by emonddr** [comment]
@dhmlau , my recent docs PR https://github.com/strongloop/loopback-next/pull/2977 only covers adding the @authenticate decorator at the method level.
